### PR TITLE
HSEARCH-3998 Mass indexing a (non-abstract) parent class in a type hierarchy with a subclass annotated with @Indexed(enabled = false) still indexes that subclass

### PR DESCRIPTION
**Pull request by yrodiere**
https://hibernate.atlassian.net/browse/HSEARCH-3998

### CODE REVIEW

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/massindexing/impl/MassIndexingIndexedTypeGroup.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	 * @return One or more type groups that are guaranteed to be disjoint.
+	 */
+	public static List<MassIndexingIndexedTypeGroup<?>> disjoint(
+			Set<? extends HibernateOrmMassIndexingIndexedTypeContext<?>> indexedTypeContexts) {
+		List<MassIndexingIndexedTypeGroup<?>> typeGroups = new ArrayList<>();
+		for ( HibernateOrmMassIndexingIndexedTypeContext<?> typeContext : indexedTypeContexts ) {
+			MassIndexingIndexedTypeGroup<?> typeGroup = MassIndexingIndexedTypeGroup.single( typeContext );
+			// First try to merge this new type group with an existing one
+			ListIterator<MassIndexingIndexedTypeGroup<?>> iterator = typeGroups.listIterator();
+			while ( iterator.hasNext() ) {
```
Comment (fax4ever) [inline mapper/orm/src/main/java/org/hibernate/search/mapper/orm/massindexing/impl/MassIndexingIndexedTypeGroup.java:43]: I don't know if we could use a `Union-Find` here to improve the complexity. See https://en.wikipedia.org/wiki/Disjoint-set_data_structure.

**Review by fax4ever — APPROVED** [review]
Very nice job! Just a comment.
+1 to merge

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/massindexing/impl/MassIndexingIndexedTypeGroup.java (hunk 1 above)
Comment (yrodiere) [reply mapper/orm/src/main/java/org/hibernate/search/mapper/orm/massindexing/impl/MassIndexingIndexedTypeGroup.java:43]: I don't know either. This code is less than 20 lines long though, so I'd say let's merge it and get back to it if we ever come by a library that provides an on-the-shelf algorithm that does exactly what we want...

Comment (yrodiere) [reply mapper/orm/src/main/java/org/hibernate/search/mapper/orm/massindexing/impl/MassIndexingIndexedTypeGroup.java:43]: Oh, you meant time complexity. Yeah, I don't think it's that important here. It's only executed once before mass indexing, and worst case we'll have what? 200 entity types? Even if this takes one second of CPU, mass indexing will take much, much longer due to heavy I/O. Not worth spending time on, at least not until we actually see a problem in performance tests.

**Comment by yrodiere** [comment]
Merged, thanks

**Pull request merged**